In [1]:
#Aiman Fiaz
#Professor Erik
#Assignment 03

import time
import math

#naive summation
def naive_sum(N): 
    sum, value = 0, math.sqrt(23)
    for i in range(N):
        sum = sum + value
    return sum

#compensated summation
def comp_sum(N):
    s, value, e = 0, math.sqrt(23), 0
    for i in range(N):
        temp = s
        y = value + e
        s = temp + y
        e = (temp - s) + y
    return s + e

#different cases of N
for N in [1000, 10000, 100000, 1000000, 10000000, 100000000]: 
    exact_sum = N*math.sqrt(23) #exact value of the sum using different N cases
    print('N =', N, '\n')
    print('Exact Sum:', exact_sum, '\n')
    
#naive summation
    start_time = time.time() # start time
    naive = naive_sum(N) # compute naive sum
    end_time = time.time() # end time
    print('Naive Sum:', naive)
    print('Absolute Error:', abs(naive-exact_sum))#formula for absolute error
    print('Relative Error:', abs((naive-exact_sum)/exact_sum))#formula for relative error
    print('Elapsed Time:', (end_time-start_time), '\n')#formula for elapsed time
    
#compensated summation
    start_time = time.time() # start time
    comp = comp_sum(N) # compute  sum
    end_time = time.time() # end time
    print('Compensated Sum:', comp)
    print('Absolute Error:', abs(comp-exact_sum))#formula for absolute error
    print('Relative Error:', abs((comp-exact_sum)/exact_sum))#formula for relative error
    print('Elapsed Time:', (end_time-start_time))#formula for elapsed time
    print('\n')

N = 1000 

Exact Sum: 4795.831523312719 

Naive Sum: 4795.831523312701
Absolute Error: 1.7280399333685637e-11
Relative Error: 3.603212341735726e-15
Elapsed Time: 6.198883056640625e-05 

Compensated Sum: 4795.831523312719
Absolute Error: 0.0
Relative Error: 0.0
Elapsed Time: 0.00014066696166992188


N = 10000 

Exact Sum: 47958.31523312719 

Naive Sum: 47958.315233122565
Absolute Error: 4.627509042620659e-09
Relative Error: 9.649023365658618e-14
Elapsed Time: 0.0005528926849365234 

Compensated Sum: 47958.31523312719
Absolute Error: 0.0
Relative Error: 0.0
Elapsed Time: 0.0008268356323242188


N = 100000 

Exact Sum: 479583.1523312719 

Naive Sum: 479583.1523311837
Absolute Error: 8.818460628390312e-08
Relative Error: 1.8387761508141895e-13
Elapsed Time: 0.0044400691986083984 

Compensated Sum: 479583.1523312719
Absolute Error: 0.0
Relative Error: 0.0
Elapsed Time: 0.010235786437988281


N = 1000000 

Exact Sum: 4795831.52331272 

Naive Sum: 4795831.523305233
Absolute Error: 7.486902177

## Analysis and Discussion

#This program calculates Sn using two different methods which we learned in class (Niave summation and Compensated summation)
#We tested 6 different values of N (10^3, 10^4, 10^5, 10^6, 10^7, and 10^8) and a comparison is made between the results of both methods with the exact sum calculated by using the formula provided in the assignment (N*math.sqrt(23)).
#As we can see in the result that Elapsed time for compensated summation is a little slower than the Elapsed time for naive summation. Which is true becuase there are more things happening in the Compensated summation than in Naive summation.
#We can also see that Compensated summation is more accurate than Naive summation. Also, the Absolute error and Relative error is really small for Niave summation.
#Elapsed time changes every time we run the program. This is because there can be other things running in the background which can slow the process.
#The answer for compensated sum is the same as the answer for Exact sum, this is why the absolute and relative error for Compensated summation is equal to 0.


#Table

#   N                 1000                     10000                  100000                  1000000                 10000000                100000000

#   Exact sum         4795.831523312719        47958.31523312719      479583.1523312719       4795831.52331272        47958315.23312719       479583152.3312719

#   Naive sum         4795.831523312701        47958.315233122565     479583.1523311837       4795831.523305233       47958315.23712847       479583151.50250727

#   Absolute Error    1.7280399333685637e-11   4.627509042620659e-09  8.818460628390312e-08   7.486902177333832e-06   0.0040012747049331665   0.8287646174430847
#  (for naive sum)

#   Realtive Error    3.603212341735726e-15    9.649023365658618e-14  1.8387761508141895e-13  1.5611270206094012e-12  8.34323450580534e-11    1.7280936859738056e-09
#  (for naive sum)

#   Elapsed Time      0.00012111663818359375   0.0007810592651367188  0.012231111526489258    0.04171299934387207     0.40421414375305176     4.040650844573975
#  (for naive sum)

#   Compensated sum   4795.831523312719        47958.31523312719      479583.1523312719       4795831.52331272        47958315.23312719       479583152.3312719

#   Absolute Error     0.0                      0.0                   0.0                      0.0                     0.0                    0.0
# (for compensated sum) 

#   Realtive Error     0.0                      0.0                   0.0                      0.0                     0.0                    0.0
# (for compensated sum) 

#   Elapsed Time       0.00015211105346679688  0.0029532909393310547  0.017965078353881836    0.0880880355834961      0.9753892421722412      9.151679039001465

